File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public int strStr(String haystack, String needle) {
        // Handle the edge case where needle is an empty string
        if (needle.isEmpty()) {
            return 0;
        }

        int haystackLength = haystack.length();
        int needleLength = needle.length();

        // Iterate through the haystack, ensuring we do not exceed the bounds
        for (int i = 0; i <= haystackLength - needleLength; i++) {
            // Check if the substring matches the needle
            if (haystack.substring(i, i + needleLength).equals(needle)) {
                return i;
            }
        }

        // Return -1 if the needle is not found
        return -1;
    }
}

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        l, r = 0, len(needle) - 1

        while l <= len(haystack) - len(needle):
            if haystack[l:].startswith(needle):
                return l
            else:
                l += 1
        
        return -1